File: packages/fx-core/templates/plugins/resource/aad/auth/bot/ts/sso/ssoDialog.ts
import {
  ComponentDialog,
  WaterfallDialog,
  Dialog,
  DialogTurnResult,
  DialogSet,
  DialogTurnStatus,
} from "botbuilder-dialogs";
import { ActivityTypes, Storage, tokenExchangeOperationName, TurnContext } from "botbuilder";
import { TeamsBotSsoPrompt, TeamsBotSsoPromptTokenResponse, TeamsFx } from "@microsoft/teamsfx";
import "isomorphic-fetch";

const DIALOG_NAME = "SSODialog";
const TEAMS_SSO_PROMPT_ID = "TeamsFxSsoPrompt";
const COMMAND_ROUTE_DIALOG = "CommandRouteDialog";

export class SsoDialog extends ComponentDialog {
  private requiredScopes: string[];
  private dedupStorage: Storage;
  private dedupStorageKeys: string[] = [];
  private commandMapping: Map<string, string | RegExp> = new Map<string, string | RegExp>();

  constructor(dedupStorage: Storage, requiredScopes: string[]) {
    super(DIALOG_NAME);

    this.initialDialogId = COMMAND_ROUTE_DIALOG;

    this.dedupStorage = dedupStorage;
    this.dedupStorageKeys = [];
    this.requiredScopes = requiredScopes;

    const teamsFx = new TeamsFx();
    const ssoDialog = new TeamsBotSsoPrompt(teamsFx, TEAMS_SSO_PROMPT_ID, {
      scopes: this.requiredScopes,
      endOnInvalidMessage: true,
    });
    this.addDialog(ssoDialog);

    const commandRouteDialog = new WaterfallDialog(COMMAND_ROUTE_DIALOG, [
      this.commandRouteStep.bind(this),
    ]);
    this.addDialog(commandRouteDialog);
  }

  public addCommand(
    commandId: string,
    commandText: string | RegExp,
    operation: (context: TurnContext, ssoToken: string) => Promise<DialogTurnResult>
  ): void {
    const dialog = new WaterfallDialog(commandId, [
      this.ssoStep.bind(this),
      this.dedupStep.bind(this),
      async (stepContext: any) => {
        const tokenResponse: TeamsBotSsoPromptTokenResponse = stepContext.result;
        const context: TurnContext = stepContext.context;
        if (tokenResponse) {
          await operation(context, tokenResponse.ssoToken);
        } else {
          await context.sendActivity("Failed to retrieve user token from conversation context.");
        }
        return await stepContext.endDialog();
      },
    ]);

    if (this.commandMapping.has(commandId)) {
      throw new Error(`Cannot add command. There is already a command with same id ${commandId}`);
    }
    this.commandMapping.set(commandId, commandText);
    this.addDialog(dialog);
  }

  /**
   * The run method handles the incoming activity (in the form of a DialogContext) and passes it through the dialog system.
   * If no dialog is active, it will start the default dialog.
   * @param {*} dialogContext
   */
  public async run(context: TurnContext, accessor: any) {
    const dialogSet = new DialogSet(accessor);
    dialogSet.add(this);

    const dialogContext = await dialogSet.createContext(context);
    const results = await dialogContext.continueDialog();
    if (results && results.status === DialogTurnStatus.empty) {
      await dialogContext.beginDialog(this.id);
    }
  }

  private async commandRouteStep(stepContext: any) {
    const turnContext = stepContext.context as TurnContext;

    // remove the mention of this bot
    let text = TurnContext.removeRecipientMention(turnContext.activity);
    if (text) {
      // remove the line break
      text = text.toLowerCase().replace(/\n|\r/g, "").trim();
    }

    const commandId = this.matchCommands(text);
    if (commandId) {
      return await stepContext.beginDialog(commandId);
    }
    return await stepContext.endDialog();
  }

  private async ssoStep(stepContext: any) {
    return await stepContext.beginDialog(TEAMS_SSO_PROMPT_ID);
  }

  private async dedupStep(stepContext: any) {
    const tokenResponse = stepContext.result;
    // Only dedup after ssoStep to make sure that all Teams client would receive the login request
    if (tokenResponse && (await this.shouldDedup(stepContext.context))) {
      return Dialog.EndOfTurn;
    }
    return await stepContext.next(tokenResponse);
  }

  public async onEndDialog(context: TurnContext) {
    const conversationId = context.activity.conversation.id;
    const currentDedupKeys = this.dedupStorageKeys.filter((key) => key.indexOf(conversationId) > 0);
    await this.dedupStorage.delete(currentDedupKeys);
    this.dedupStorageKeys = this.dedupStorageKeys.filter((key) => key.indexOf(conversationId) < 0);
  }

  // If a user is signed into multiple Teams clients, the Bot might receive a "signin/tokenExchange" from each client.
  // Each token exchange request for a specific user login will have an identical activity.value.Id.
  // Only one of these token exchange requests should be processed by the bot. For a distributed bot in production,
  // this requires a distributed storage to ensure only one token exchange is processed.
  private async shouldDedup(context: TurnContext): Promise<boolean> {
    const storeItem = {
      eTag: context.activity.value.id,
    };

    const key = this.getStorageKey(context);
    const storeItems = { [key]: storeItem };

    try {
      await this.dedupStorage.write(storeItems);
      this.dedupStorageKeys.push(key);
    } catch (err) {
      if (err instanceof Error && err.message.indexOf("eTag conflict")) {
        return true;
      }
      throw err;
    }
    return false;
  }

  private getStorageKey(context: TurnContext): string {
    if (!context || !context.activity || !context.activity.conversation) {
      throw new Error("Invalid context, can not get storage key!");
    }
    const activity = context.activity;
    const channelId = activity.channelId;
    const conversationId = activity.conversation.id;
    if (activity.type !== ActivityTypes.Invoke || activity.name !== tokenExchangeOperationName) {
      throw new Error("TokenExchangeState can only be used with Invokes of signin/tokenExchange.");
    }
    const value = activity.value;
    if (!value || !value.id) {
      throw new Error("Invalid signin/tokenExchange. Missing activity.value.id.");
    }
    return `${channelId}/${conversationId}/${value.id}`;
  }

  private matchCommands(text: string): string {
    for (const command of this.commandMapping) {
      const pattern: string | RegExp = command[1];
      let matchResult: RegExpExecArray | boolean;
      if (typeof pattern == "string") {
        matchResult = text === pattern;
      } else {
        matchResult = pattern.exec(text);
      }
      if (matchResult) {
        return command[0]; // Return the command id
      }
    }

    return undefined;
  }
}

File: packages/fx-core/templates/plugins/resource/aad/auth/bot/csharp/SSO/SsoDialog.cs
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Options;
using Microsoft.TeamsFx.Bot;
using Microsoft.TeamsFx.Configuration;

namespace {Your_NameSpace}.SSO;

public class DialogConstants
{
    internal static readonly string COMMAND_ROUTE_DIALOG = "CommandRouteDialog";
    internal static readonly string TEAMS_SSO_PROMPT = "TeamsFxSsoPrompt";
}

public class StoreItem
{
    public string eTag;
}

public class SsoDialog : ComponentDialog
{
    ILogger<SsoDialog> _logger;
    BotAuthenticationOptions _botAuthOptions;
    Dictionary<string, string> _commandMapping = new Dictionary<string, string>();
    IStorage _dedupStorage = new MemoryStorage();
    List<string> _dedupStorageKeys = new List<string>() {};

    public SsoDialog(IOptions<BotAuthenticationOptions> botAuthenticationOptions, ILogger<SsoDialog> logger)
    {
        _logger = logger;
        InitialDialogId = DialogConstants.COMMAND_ROUTE_DIALOG;

        try
        {
            _logger.LogTrace("Validate bot authentication configuration");
            _botAuthOptions = botAuthenticationOptions.Value;
        }
        catch (OptionsValidationException e)
        {
            throw new Exception($"Bot authentication config is missing or not correct with error: {e.Message}");
        }

        var settings = new TeamsBotSsoPromptSettings(_botAuthOptions, new string[] { "User.Read" });
        AddDialog(new TeamsBotSsoPrompt(DialogConstants.TEAMS_SSO_PROMPT, settings));

        WaterfallDialog commandRouteDialog = new WaterfallDialog(
            DialogConstants.COMMAND_ROUTE_DIALOG,
            new WaterfallStep[]
            {
                CommandRouteStepAsync
            });
        AddDialog(commandRouteDialog);

        _logger.LogInformation("Construct Main Dialog");
    }

    public async Task RunAsync(ITurnContext context, IStatePropertyAccessor<DialogState> accessor)
    {
        DialogSet dialogSet = new DialogSet(accessor);
        dialogSet.Add(this);

        DialogContext dialogContext = await dialogSet.CreateContextAsync(context);
        DialogTurnResult results = await dialogContext.ContinueDialogAsync();
        if (results != null && results.Status == DialogTurnStatus.Empty)
        {
            await dialogContext.BeginDialogAsync(Id);
        }
    }

    public void addCommand(
        string commandId,
        string commandText,
        Func<ITurnContext, string, BotAuthenticationOptions, Task> operation)
    {
        if (_commandMapping.ContainsValue(commandId))
        {
            return;
        }
        _commandMapping.Add(commandText, commandId);

        WaterfallDialog dialog = new WaterfallDialog(
            commandId,
            new WaterfallStep[]
            {
            PromptStepAsync,
            DedupStepAsync,
            async (WaterfallStepContext stepContext, CancellationToken cancellationToken) => {
                TeamsBotSsoPromptTokenResponse tokenResponce = (TeamsBotSsoPromptTokenResponse)stepContext.Result;
                var turnContext = stepContext.Context;
                try
                {
                    if (tokenResponce != null)
                    {
                        await operation(turnContext, tokenResponce.Token, _botAuthOptions);
                    } else
                    {
                        await turnContext.SendActivityAsync("Failed to retrieve user token from conversation context.");
                    }
                    return await stepContext.EndDialogAsync();
                } catch (Exception error)
                {
                    await turnContext.SendActivityAsync("Failed to retrieve user token from conversation context.");
                    await turnContext.SendActivityAsync(error.Message);
                    return await stepContext.EndDialogAsync();
                }
            }
        });
        AddDialog(dialog);
    }

    protected override async Task OnEndDialogAsync(ITurnContext context, DialogInstance instance, DialogReason reason, CancellationToken cancellationToken = default(CancellationToken))
    {
        var conversationId = context.Activity.Conversation.Id;
        var currentDedupKeys = _dedupStorageKeys.Where((key) => key.IndexOf(conversationId) > 0).ToArray();
        await _dedupStorage.DeleteAsync(currentDedupKeys);
        _dedupStorageKeys = _dedupStorageKeys.Where((key) => key.IndexOf(conversationId) < 0).ToList<string>();
    }

    private async Task<DialogTurnResult> PromptStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
    {
        _logger.LogInformation("Step: Prompt to get SSO token");
        try
        {
            return await stepContext.BeginDialogAsync(DialogConstants.TEAMS_SSO_PROMPT, null, cancellationToken);
        } catch (Exception error)
        {
            await stepContext.Context.SendActivityAsync("Failed to run SSO prompt");
            await stepContext.Context.SendActivityAsync(error.Message);
            return await stepContext.EndDialogAsync();
        }
    }

    private async Task<DialogTurnResult> CommandRouteStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
    {
        _logger.LogInformation("Step: Route to pre-added commands");
        var turnContext = stepContext.Context;

        string text = turnContext.Activity.RemoveRecipientMention();
        if (text != null)
        {
            text = text.ToLower().Trim();
        }

        string commandId = MatchCommands(text);
        if (commandId != null)
        {
            return await stepContext.BeginDialogAsync(commandId);
        }

        await stepContext.Context.SendActivityAsync(String.Format("Cannot find command: {0}", text));
        return await stepContext.EndDialogAsync();
    }

    private async Task<DialogTurnResult> DedupStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
    {
        try
        {
            var tokenResponse = stepContext.Result;
            if (tokenResponse != null && (await ShouldDedup(stepContext.Context)))
            {
                return EndOfTurn;
            }
            return await stepContext.NextAsync(tokenResponse);
        }
        catch (Exception error)
        {
            await stepContext.Context.SendActivityAsync("Failed to run dedup step");
            await stepContext.Context.SendActivityAsync(error.Message);
            return await stepContext.EndDialogAsync();
        }
    }

    private async Task<bool> ShouldDedup(ITurnContext context)
    {
        var storeItem = new StoreItem()
        {
            eTag = (context.Activity.Value as dynamic).id,
        };
        var key = GetStorageKey(context);
        var storeItems = new Dictionary<string, object>()
        {
            {key, storeItem}
        };

        var res = await _dedupStorage.ReadAsync(new string[] { key });
        if (res.Count != 0)
        {
            return true;
        }

        await _dedupStorage.WriteAsync(storeItems);
        _dedupStorageKeys.Add(key);
        return false;
    }

    private string GetStorageKey(ITurnContext context)
    {
        if (context == null || context.Activity == null || context.Activity.Conversation == null)
        {
            throw new Exception("Invalid context, can not get storage key!");
        }

        var activity = context.Activity;
        var channelId = activity.ChannelId;
        var conversationId = activity.Conversation.Id;
        if (activity.Type != ActivityTypes.Invoke || activity.Name != SignInConstants.TokenExchangeOperationName)
        {
            throw new Exception("TokenExchangeState can only be used with Invokes of signin/tokenExchange.");
        }

        var value = activity.Value;
        if (value == null || (value as dynamic).id == null)
        {
            throw new Exception("Invalid signin/tokenExchange. Missing activity.value.id.");
        }

        return $"{channelId}/{conversationId}/{(value as dynamic).id}";
    }

    private string MatchCommands(string text)
    {
        if (_commandMapping.ContainsKey(text))
        {
            return _commandMapping[text];
        }

        return null;
    }
}
